### PR TITLE
Comments

### PR DESCRIPTION
**Pull request by YiFeiZhang2**
Add comments for middleware and validator functions.
Add apiParams for current routes.
Move constants to constants.js, and add comments for url regexes.
Simplify mongoIDArray validator logic by moving the custom code to another function.

### CODE REVIEW

**Comment by pavelkondratyev** [comment]
Approving for when you fix the check, looks good otherwise

File: middlewares/validators/validator.helper.js [hunk 1]
```diff
@@ -86,32 +112,22 @@ function mongoIdArrayValidator (getOrPost, fieldname, optional = true) {
...
-                    }
-                });
-                return true;
-            }).withMessage("Each element must be a valid mongoId");
+                return isMongoIdArray(value);
+            }).withMessage("Value must be an array of mongoIDs");
     } else {
-        return arr.exists()
+            return arr.exists()
             .custom((value) => {
```
[the remark this answers is not in the document]
Comment (YiFeiZhang2) [reply middlewares/validators/validator.helper.js]: Done

File: middlewares/validators/validator.helper.js [hunk 2]
```diff
@@ -294,14 +359,24 @@ function applicationValidator (getOrPost, fieldname, optional = true) {
...
+ * @param {boolean} optional Whether the field is optional or not.
+ */
+function isMongoIdArray (arr) {
+    if (!Array.isArray(arr)) {
+        return false;
+    }
+
+    arr.forEach(ele => {
+        if (!mongoose.Types.ObjectId.isValid(ele)) {
+            return false;
```
[the remark this answers is not in the document]
Comment (YiFeiZhang2) [reply middlewares/validators/validator.helper.js]: You're right. Fixed.

File: routes/api/hacker.js [hunk 3]
```diff
@@ -28,7 +28,24 @@ module.exports = {
          * @apiParam (body) {String} school Name of the school the hacker goes to
          * @apiParam (body) {String} gender Gender of the hacker
          * @apiParam (body) {Boolean} needsBus Whether the hacker requires a bus for transportation
-         * @apiParam (body) {Object} application The hacker's application
+         * @apiParam (body) {Object} application The hacker's application. 
+         * An example application:
```
[the remark this answers is not in the document]
Comment (YiFeiZhang2) [reply routes/api/hacker.js]: Done.

**Pull request merged**